Bugfix/fix intrinsic header for arm architecture

minosgalanakis opened the pull request:
## Description

Currently the development branch fails to compile in arm silicone because of how the arm intrisics macros are not properly defined. This pr changes the inclusion order of `tf_psa_crypto_common.h` after they have been defined, in affected source files.

It also updates to the head of framework to bring in other related bugfixes in the tf-psa-crypto branch. 

## PR checklist

Please remove the segment/s on either side of the | symbol as appropriate, and add any relevant link/s to the end of the line.
If the provided content is part of the present PR remove the # symbol.

- [x] **changelog** not required because: the bug that is fixed by this PR has not been released
- [x] **framework PR** not required
- [x] **mbedtls development PR** not required because: The crypto files have been moved to TF-PSA-Crypto
- [x] **mbedtls 3.6 PR** provided Mbed-TLS/mbedtls#10542
- **tests**  provided | not required because:

ronald-cron-arm commented:
If you base your PR on current development, you won't need the framework pointer update. It is already there.

gilles-peskine-arm commented on drivers/builtin/src/aesce.c:
```diff
@@ -5,10 +5,6 @@
 
-#include "tf_psa_crypto_common.h"
-
-#if defined(MBEDTLS_AESCE_C)
```
The fix for `aesce.c` is already approved in https://github.com/Mbed-TLS/TF-PSA-Crypto/pull/603.

But I'd missed that `sha256.c` was also affected, thanks.

minosgalanakis commented:
Latest force [push](https://github.com/Mbed-TLS/TF-PSA-Crypto/compare/3429951e008f6df1a4190b8fff620fa03dcfe184..d17897bac6e9c4aa5d570f1f9d07011b263ce2d8) was a rebase to latest development.

ronald-cron-arm commented on drivers/builtin/src/sha512.c:
```diff
@@ -24,7 +24,10 @@
  *
  * `arm_neon.h` is included by tf_psa_crypto_common.h, so we put these defines
- * at the top of this file, before any includes.
+ * at the top of this file, before any includes. This is necessary with
```
It seems we have the same issue in sha512.c.

ronald-cron-arm commented on drivers/builtin/src/sha512.c:
```diff
@@ -24,7 +24,10 @@
  *
  * `arm_neon.h` is included by tf_psa_crypto_common.h, so we put these defines
- * at the top of this file, before any includes.
+ * at the top of this file, before any includes. This is necessary with
```
"at the top of this file, before any includes."

That's not true. `#include "tf_psa_crypto_common.h"` is before.

gilles-peskine-arm replied:
But it shouldn't be, that's the point.

gilles-peskine-arm commented on drivers/builtin/src/sha512.c:
Here as well, the `#include` directive needs to be moved back to after `#define __ARM_FEATURE_…`

minosgalanakis replied:
Done as of https://github.com/Mbed-TLS/TF-PSA-Crypto/pull/614/commits/7172c4aa6e04bb722e731c483eee9dca29f5c91d

gilles-peskine-arm reviewed (changes requested):
`sha512.c` has the same problem as `sha256.c`.

Maybe a better fix would be to move the `#define __ARM_FEATURE_xxx` definitions to `tf_psa_crypto_common.h` (`common.h` in 3.6), before it includes `<arm_neon.h>`? That would seem like the natural approach to me, but I don't know why it wasn't done that way originally.

minosgalanakis commented:
> `sha512.c` has the same problem as `sha256.c`.
> 
> Maybe a better fix would be to move the `#define __ARM_FEATURE_xxx` definitions to `tf_psa_crypto_common.h` (`common.h` in 3.6), before it includes `<arm_neon.h>`? That would seem like the natural approach to me, but I don't know why it wasn't done that way originally.

I have moved the include in sha512 for now and rebase on top of latest development to pick up your changes. I am preparing a patch to test moving it to tf_psa_crypto_common but I could not get it to work on my computer as of now. I will update the pr as soon as its done.

minosgalanakis commented:
I have done another re-base and no other changes since the previous one. Changes https://github.com/Mbed-TLS/TF-PSA-Crypto/compare/3aa58368936dde5001015db92a1d8a22c7bfa478..577f51c846987b786601c42adecf5791565976b5 are rebased content. 

I will be following up with some changes as requested above

minosgalanakis commented:
Push to https://github.com/Mbed-TLS/TF-PSA-Crypto/commit/e48c6ec38c433e30696a34f5b079f93473fd44b1 is a simple rebase.

gilles-peskine-arm reviewed (dismissed):
LGTM

ronald-cron-arm reviewed (dismissed):
LGTM

minosgalanakis commented:
Added commit https://github.com/Mbed-TLS/TF-PSA-Crypto/pull/614/commits/a1d8494067c904b44c5325991c9b9b7b141436eb which is implementing what @gilles-peskine-arm was asking at https://github.com/Mbed-TLS/TF-PSA-Crypto/pull/614#pullrequestreview-3605551266. 

I clean-up the the include logic, but this commit is a quality of life improvement and I could undo it if you do not feel it is necessary.

minosgalanakis committed:
drivers sha256|512: Adjusted tf_psa_crypto_common.h inclusion

This patch adjusts the include order so that some ACLE
intrinsics macros are configured before
the inclusion of `neon.h`. This fixes issues with older
clang compilers but has no effect in modern versions.

Signed-off-by: Minos Galanakis <minos.galanakis@arm.com>

minosgalanakis commented:
Latest head is identical to the previously approved head -> [link ](https://github.com/Mbed-TLS/TF-PSA-Crypto/commit/c9a17193020cdda59109b2a1e57b946ad1a8b4e2) . You can check the  comparison -> [link](https://github.com/Mbed-TLS/TF-PSA-Crypto/compare/c9a17193020cdda59109b2a1e57b946ad1a8b4e2..c8f1f771ae61079b70955f8b661d6c612afedd93)

The proposed change by https://github.com/Mbed-TLS/TF-PSA-Crypto/commit/93153586af016cf7316d625c754db94d2fa3e792 will be moved to a separate pr because it needs a design-aproved change on how we handle `check-names.py`

I also cleaned up the commit stack since the change is minimal.

ronald-cron-arm reviewed (approved):
LGTM

gilles-peskine-arm reviewed (approved):
LGTM (identical content to c9a17193020cdda59109b2a1e57b946ad1a8b4e2 which I previously approved)
